woocommerce/woocommerce-gateway-stripe
# Filter payment methods at checkout by active capability status

# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #1917 

Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.

Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.

> With Stripe payment methods, they are enabled on Stripe with capabilities. If a payment method that has not been properly enabled on Stripe is added to the UPE, UPE will throw an error. To avoid this scenario we need to filter out any payment methods from the UPE based on their capability status. 
>
> In short, we can get the capabilities status from the Stripe GET /account endpoint. And if the payment method capability has an active status, it can be used in UPE, anything other than active and it should not be used in UPE.

I took advantage of the existing `WC_Stripe_Account`, which manages a cached transient containing the `/account` response from Stripe. The class handles the request and checks to ensure all required account keys are provided, so this seemed like a good idea.

Another caveat is that the way this check is designed, if for some reason we are unable to get a capabilities response from Stripe, we would end up disabling all of the payment methods. This might be desired, but it leaves the checkout incredibly broken and without really alerting the user that something isn't right. To mitigate this, I decided to set the credit card payment method to always return an affirmative value when it's capability activity is checked, so that at least this payment method will be present at all times. Let me know if this is desirable or not!

# Testing instructions

Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?

1. Please pull the changes in this PR, ensure the UPE is enabled, and that valid test keys are stored in settings (of course).
2. Visit checkout and confirm that payment methods are present.
3. Toggle capabilities `active` status from the Stripe dashboard or via [API request](https://stripe.com/docs/api/capabilities/update). Check that payment method presence reflects the `active` status.

P.S. I still need to test this a little bit more, as I'm not totally clear how capabilities work in test mode, but I'd like to get this PR out there anyways, because there are a couple weird things that I found along the way.

-------------------

(remove me)
